akka-cluster/src/main/scala/akka/cluster/JoinConfigCompatChecker.scala

```diff
@@ -0,0 +1,166 @@
+package akka.cluster
+
+import java.util
+
+import akka.actor.{ ActorSystem, ExtendedActorSystem }
+import akka.annotation.InternalApi
+import com.typesafe.config.{ Config, ConfigFactory, ConfigUtil, ConfigValue }
+
+import scala.collection.JavaConverters._
+import scala.collection.{ immutable ⇒ im }
+
+trait JoinConfigCompatChecker {
+
+  /** A Set of configuration keys that are required for this checker */
+  def requiredKeys: Set[String]
+
+  /**
+   * Runs the Config check.
+   *
+   * Implementers are free to define what makes Config entry compatible or not.
+   * We do provide some pre-build checks tough: [[JoinConfigCompatChecker.exists()]] and [[JoinConfigCompatChecker.fullMatch()]]
+   *
+   *
+   * @param toCheck - the Config instance to be checked
+   * @param currentConfig - the Config instance containing the expected values
+   * @return a [[ConfigCheck]]. Can be [[Compatible]] or [[Incompatible]], the later must contain a descriptive list of error messages.
+   */
+  def check(toCheck: Config, currentConfig: Config): ConfigCheck
+}
+
+object JoinConfigCompatChecker {
+
+  /**
+   * Checks that all `requiredKeys` are available in `toCheck` Config.
+   *
+   * @param requiredKeys - a Set of required keys
+   * @param toCheck - the Config instance to be checked
+   */
+  def exists(requiredKeys: Set[String], toCheck: Config): ConfigCheck = {
+    val allKeys = toCheck.entrySet().asScala.map(_.getKey)
+    // return all not found required keys
+    val result =
+      requiredKeys.collect {
+        case requiredKey if !allKeys.contains(requiredKey) ⇒ requiredKey + " is missing"
+      }
+
+    if (result.isEmpty) Compatible
+    else Incompatible(result.to[im.Seq])
+  }
+
+  /**
+   * Checks that all `requiredKeys` are available in `toCheck` Config
+   * and its values match exactly the values in `currentConfig`.
+   *
+   * @param requiredKeys - a Set of required keys
+   * @param toCheck - the Config instance to be checked
+   * @param currentConfig - the Config instance containing the expected values
+   */
+  def fullMatch(requiredKeys: Set[String], toCheck: Config, currentConfig: Config): ConfigCheck = {
+
+    def checkEquality = {
+
+      def checkCompat(entry: util.Map.Entry[String, ConfigValue]) = {
+        val key = entry.getKey
+        currentConfig.hasPath(key) && currentConfig.getValue(key) == entry.getValue
+      }
+
+      def buildMessage(entry: util.Map.Entry[String, ConfigValue]) = {
+        val key = entry.getKey
+        if (currentConfig.hasPath(key))
+          s"${entry.getKey} is incompatible: found '${entry.getValue.unwrapped()}', expected '${currentConfig.getValue(entry.getKey).unwrapped()}'"
```

Comment:
This also relates to previous discussion about sensitive data. Here I'm generating messages containing the found and expected values. This messages will later be used to generate a log on the receiving node.

That means that config value will appear in the logs. At first, my thought was that this is user responsibility. 

On our side, we must make sure that the checkers that we provide won't print any sensitive info. I'm not even sure that there are sensitive information in akka configs. 

In case we have to check something that we believe is sensitive we should not use this helper (`JoinConfigCompatChecker.fullMatch`) or have a way to tweak this error message. 

On the other hand, it's very useful for the users if we provide a descriptive message for each incompatible configuration.

WDYT?

Reply:
I agree that we should not log the config values, logging the config keys are probably enough.